addons/mail/static/src/messaging/entity_manager.js

```diff
@@ -0,0 +1,749 @@
+const { many2many, many2one, one2many, one2one } = require('mail.messaging.entityUtils');
+const { patchClassMethods, patchInstanceMethods } = require('mail.messaging.utils');
+
+/**
+ * Object that manage entities, notably their update cycle: whenever some
+ * entities are requested for update (either with static method `create()` or
+ * instance method `update()`), this object process them with direct field &
+ * and computed field updates.
+ */
+class EntityManager {
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    constructor(env) {
+        /**
+         * Contains all entity classes. key is name of entity name, and value
+         * is class.
+         */
+        this.classes = {};
+        /**
+         * The messaging env.
+         */
+        this.env = env;
+
+        /**
+         * Contains all entity instances. key is local id, while value is
+         * instance.
+         */
+        this._instances = {};
+        /**
+         * Whether this is currently handling an "update after" on an entity.
+         * Useful to determine if we should process computed/related fields.
+         */
+        this._isHandlingToUpdateAfters = false;
+        /**
+         * Determine whether an update cycle is currently in progress.
+         * Useful to determine whether an update should initiate an update
+         * cycle or not. An update cycle basically prioritizes processing
+         * of all direct updates (i.e. explicit from `data`) before
+         * processing computes.
+         */
+        this._isInUpdateCycle = false;
+        /**
+         * Fields flagged to call compute during an update cycle.
+         * For instance, when a field with dependents got update, dependent
+         * fields should update themselves by invoking compute at end of
+         * update cycle. Key is of format <entity-local-id>--<fieldName>, and
+         * determine entity and field to be computed.
+         */
+        this._toComputeFields = new Map();
+        /**
+         * List of "update after" on entities that have been registered.
+         * These are processed after any explicit update and computed/related
+         * fields.
+         */
+        this._toUpdateAfters = [];
+    }
+
+    /**
+     * Called when all JS modules that register or patch entities have been
+     * done. This launches generation of entity classes.
+     */
+    start() {
+        /**
+         * Generate the entities.
+         */
+        this.classes = this._generateClasses();
+        /**
+         * Make environment accessible from Entities.
+         */
+        for (const Entity of Object.values(this.classes)) {
+            Entity.env = this.env;
+        }
+        /**
+         * Create the messaging singleton entity.
+         */
+        this.messaging = this.classes.Messaging.create();
```

Comment:
This create should still be done in messaging env.
"messaging" word should never appear in entity manager/field/core (make entity "ready" to be moved in web)